larpandoracontent/LArThreeDReco/LArEventBuilding/NeutrinoHierarchyAlgorithm.cc

```diff
@@ -214,6 +222,53 @@ void NeutrinoHierarchyAlgorithm::ProcessPfoInfoMap(const ParticleFlowObject *con
 
 //------------------------------------------------------------------------------------------------------------------------------------------
 
+void NeutrinoHierarchyAlgorithm::AdjustVertexAndPfoInfo(const ParticleFlowObject *const pNeutrinoPfo, const PfoList &candidateDaughterPfoList,
+    PfoInfoMap &pfoInfoMap) const
+{
+    PfoVector candidateDaughterPfoVector(candidateDaughterPfoList.begin(), candidateDaughterPfoList.end());
+    std::sort(candidateDaughterPfoVector.begin(), candidateDaughterPfoVector.end(), LArPfoHelper::SortByNHits);
+
+    // TODO Consider deleting the neutrino pfo if there are no daughter pfo candidates
+    if (candidateDaughterPfoVector.empty())
+        throw StatusCodeException(STATUS_CODE_NOT_FOUND);
+
+    ClusterList daughterClusterList3D;
+    LArPfoHelper::GetThreeDClusterList(candidateDaughterPfoVector.front(), daughterClusterList3D);
+
+    if (daughterClusterList3D.empty())
+        throw StatusCodeException(STATUS_CODE_NOT_FOUND);
+
+    const Vertex *pOldNeutrinoVertex(LArPfoHelper::GetVertex(pNeutrinoPfo));
+    const CartesianVector newVertexPosition(LArClusterHelper::GetClosestPosition(pOldNeutrinoVertex->GetPosition(), daughterClusterList3D.front()));
```

Comment:
Am I wrong in thinking that the newVertexPosition is semi-arbitrary because it is set to the closest distance between the old vertex and whichever 3D cluster is at the front of the unsorted daughterClusterList3D list?  Should we compare to all 3D clusters in this list instead?

Reply:
I now see in the ordering that the PFO used to get the 3D clusters is the one with the largest number of PFOs.  Is there any mechanism by which a PFO can have more than one 3D cluster at this stage?  If not then this is fine, but it might be worth adding a comment saying something along the lines of this is the largest 3D cluster in the code as it wasn't immediately obvious when reading through (although your git comment is very nice and clear).

Reply:
Hi Steve,

Thanks - you're right that this has ordered the pfos and got the first 3D cluster (which is from a list that holds its order but isn't explicitly sorted), so there's an implicit assumption that there will only be one 3D cluster per pfo.

Right now, I think every pfo will have maximum one cluster of each of the following hit types: u, v, w, 3D. I'm not aware of any that have two of a given hit type (would be due to an error), but it also wouldn't hurt to add a quick sort of the 3D clusters in this function anyhow.

Would this work for you?

Reply:
Adding a sort would make things very clear and help the readability I think, but I'll leave the final decision up to you because a comment would equally help that aspect and save having to call the albeit trivial sort.  But there's also the issue of comments going out of date and so I can see benefits either way.